### PR TITLE
feature/28618_rule_class_hierarchy: Django and the scanner should have equivalent concepts of rules

### PR DESCRIPTION
**Pull request by magenta-af**
Refactoring any of the rule code is unnecessarily difficult, as `django-os2webscanner` and `scrapy-webscanner` don't really agree on what a rule _is_.

From the Django perspective, a rule is a `RegexRule` with one or more regular expressions, unless it has optional CPR-specific settings, in which case it can have zero - or, if it's a name or address rule, it's a property on a scanner(?!). In the scanner's universe, meanwhile, a rule is a `RegexRule`, a `NameRule`, or an `AddressRule`, unless it's a `RegexRule` with CPR-specific settings, in which case `RegexRule` instantiates `CPRRule` objects and defers to them for matches.

This is, er, a bit of a mess. This branch changes that: both Django and the scanner are now aware that scanners are associated with `Rule`s, and that a `Rule` can be a `RegexRule`, a `CPRRule`, a `NameRule`, or an `AddressRule`. (The two universes still have different implementations of these concepts, because they use them in different ways, but there are no longer gaping incompatibilities between the two representations.)

(Whether we want this in release 2.2.0 or not is an open question; it is quite a big change...)

### CODE REVIEW

**Comment by magenta-af** [comment]
I see that commit 0acb47e17b7c3766a6d3c5b5d401dfe6b518cebe steps on a bunch of the database fields that this branch _also_ steps on. :-)

What's the right thing to do here? In principle, 0acb47e17b7c3766a6d3c5b5d401dfe6b518cebe is already in `development`, so I should make sure that these migrations can all be applied cleanly there, but I don't think I know how to do that; there are irreversible custom migrations in this stack...

**Comment by magenta-af** [comment]
I destroyed my database, rewound the state of the world to `development`, made some test rules, and then applied all of these changes, and... the migrations did what they should. Which was a nice surprise! Thanks for fixing them up.

**Comment by magenta-af** [comment]
With this merge migration in place, everything now seems to work as it should based on `development`. (And I'm getting really good at throwing the database away and starting from scratch! It only takes a minute!)

File: django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html [hunk 1]
```diff
@@ -9,9 +9,14 @@
   {% include 'os2webscanner/rules_above_content.html' with active='adhoc' %}
   <div class="tab-content">
     <div role="tabpanel" class="tab-pane active" id="adhoc">
-      <p class="h2" style="display:inline-block;">Ad hoc-regler</p>
+      <p class="h2" style="display:inline-block;">Regulær udtryk-regler</p>
```
Comment (danchr) [inline django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html]: IMO “Regler” is implicit given that we're on the rules page. Also, grammar:
```suggestion
      <p class="h2" style="display:inline-block;">Regulære udtryk</p>
```

File: django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html [hunk 2]
```diff
@@ -94,23 +92,141 @@
...
+            </tr>
+            {% endfor %}
+          </tbody>
+        </table>
+      </div>
+    </div>
+  </div>
+  <div class="tab-content">
+    <div role="tabpanel" class="tab-pane active" id="adhoc">
+      <p class="h2" style="display:inline-block;">CPR-regler</p>
```
Comment (danchr) [inline django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html]: Likewise, “Regler” is implicit — so:
```suggestion
      <p class="h2" style="display:inline-block;">CPR-numre</p>
```

File: django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html [hunk 3]
```diff
@@ -94,23 +92,141 @@
...
+                <strong>
+                <small>
+                Modulo-11
+                </small>
+                </strong>
+              </td>
+              <td>
+                <strong>
+                <small>
+                Fødselsdagsdato tjek
```
Comment (danchr) [inline django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html]:
```suggestion
                Kun gyldige datoer
```

File: django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html (hunk 1 above)
Comment (dals83) [reply django-os2webscanner/os2webscanner/templates/os2webscanner/rules.html]: Ad hoc regler er Mortens udtryk. Så det skal det hedde!
Den anden vil jeg godt acceptere :)

**Comment by magenta-af** [comment]
I'm... not sure! I don't see anything blocking it other than that it hasn't been reviewed yet...

@danchr, did you want me to include a migration to clean up empty `RegexRule`s in here?

**Comment by danchr** [comment]
> I'm... not sure! I don't see anything blocking it other than that it hasn't been reviewed yet...
> 
> @danchr, did you want me to include a migration to clean up empty `RegexRule`s in here?

Yeah, I think we should fix it. Although I can't say what our users have set up in actual, proper production, I don't think it's sensible to allow the _possibility_ of a leftover rule that causes widespread breakage.

File: django-os2webscanner/os2webscanner/models/rules/regexrule_model.py [hunk 4]
```diff
@@ -17,7 +17,13 @@
...
 
 from django.db import models
-from .regexrule_model import RegexRule
+
+
+from .rule_model import Rule
+
+
+class RegexRule(Rule):
+    pass
```
Comment (danchr) [inline django-os2webscanner/os2webscanner/models/rules/regexrule_model.py]: When I look at the code, this `pass` worries me a bit; especially when combined with [the line below](https://github.com/magenta-aps/os2datascanner/pull/14/files#diff-ad197e04e8247645a9f75b7384f3c46cR37):

```python
     regex = models.ForeignKey(RegexRule, null=True, on_delete=models.CASCADE,
                               related_name='patterns', verbose_name='Regex')
```

Conceptually, this relationship seems a bit backwards, and the possible cause for your current worries. I'd argue that the `pattern` forms part of the `rule`, which thus has ownership of it. Would swapping them be helpful? Also; the `RegexPattern` model currently allows a `None` regex; I don't think that makes sense.

Comment (magenta-af) [reply django-os2webscanner/os2webscanner/models/rules/regexrule_model.py]: I'm not sure I understand this issue. `RegexRule` is a `Rule`, but the only thing it has that a `Rule` doesn't have is zero (which _should_ be one!) or more `RegexPattern`s associated with it -- the `regex` `ForeignKey` in `RegexPattern` seems to be enough of a relationship for Django to create the necessary tables.

My current worries are a bit more complicated than that: `0068_create_newrules`, which introduced the basis for a concrete inheritance hierarchy without doing anything to `RegexPatterns`, and then `0069_newrule_to_rule`, which then renamed `NewRule` to `Rule` _without_ making sure that things that had pointers to _old_ `Rule`s (like `RegexPattern`s) were updated accordingly.

I have no idea how to proceed. Maybe we can just drop this branch and I'll try again another time...

**Pull request closed without merging**